Add tools to analyze the power spectral density.

chrismo-schmidt opened the pull request:
Adds two functions to analyse the power spectral density:
- pow_spectrum: Calculate the power spectral density of a signal
- cum_pow_spectrum: Calculate the cumulative power spectral density. Includes a kwarg for relative and absolute power density.

This uses the existing freq_spectrum and follows the same conventions (e.g. combine power of positive and negative frequencies in the positive frequencies). I added a "norm" keyword to freq_spectrum that makes the new numpy fft normalization keyword accessible. The power functions change the default "forward" policy (which is equal to scaling by 1/N, as previously done), to "ortho" to ensure that Parseval's theorem is satisfied.

Unittests for the two new functions and the existing freq_spectrum are included. 

An open question is why freq_spectrum removes the bias (f=0)? I think it can be quite crucial to know if a signal as a constant part.

moorepants commented on dynamicisttoolkit-dev-env.yml:
```diff
@@ -3,6 +3,7 @@ channels:
   - conda-forge
 # NOTE : The minimum versions correspond to those in Ubuntu 22.04 LTS.
 dependencies:
+  - python == 3.11
```
We need to leave this out for the CI so that it can set python. We can switch from nose to pytest so that we can run on Python 3.12 in the future. See #44

chrismo-schmidt replied:
Alright.

moorepants replied:
I fixed this in master.

moorepants commented on dtk/process.py:
```diff
@@ -525,9 +532,82 @@ def nextpow2(i):
     return frequency, amplitude
 
 
+def pow_spectrum(data, sampleRate):
```
Let's follow pep8 naming conventions for new things (for some reason I didn't follow it before.

moorepants replied:
`sample_rate`

chrismo-schmidt replied:
I fully agree. I only kept this one for consistency with ```freq_spectrum()```

moorepants commented on dtk/process.py:
```diff
@@ -513,7 +519,8 @@ def nextpow2(i):
     n = nextpow2(L)
-    Y = fft(data, n)/L  # divide by L for scaling
+
+    Y = fft(data, n, norm="forward") 
```
Does this do the same thing as the prior `/L`? Or does it change the result?

moorepants replied:
> The power functions change the default "forward" policy (which is equal to scaling by 1/N, as previously done), to "ortho" to ensure that Parseval's theorem is satisfied.

Ok, so the same.

moorepants replied:
So if no normalization occurred pre-numpy 1.10, I wonder if my implementation assumed that. The docs now say that the default is backwards, so I guess this code normalized the already backwards normalized result.

chrismo-schmidt replied:
To be an invertible transformation ( $x(n) = \mathrm{IDFT}(\mathrm{DFT}(x(n))$ ), the normalization factor $\frac{1}{N}$ can be assigned to both the forward and the backward transformation.

It is my impression that the backwards case where the normalization happens in the inverse DTF is more common, but I could be wrong. 

The old numpy did no normalization of the "forward" DFT ($X(k) = \mathrm{DFT}(x(n))$) and normalized the inverse DFT. This is exactly the default "backwards" case now. So old and new have the same default behaviour. Your code did the normalization of the forward DFT manually, effectively switching to the "forward" convention. For the spectrum, this is fine and there are good reasons to do the forward normalization. I guess the most important one is that the amplitudes are independent of the signal length. An error may only occur if someone does an default inverse DFT which would apply the normalization again (backwards).

moorepants replied:
Glad to have an electrical engineering looking at this!

moorepants commented:
> An open question is why freq_spectrum removes the bias (f=0)? I think it can be quite crucial to know if a signal as a constant part.

No idea. I believe I adopted this from a tutorial I found somewhere that explained how this worked. So it is possible it wasn't present or I decided not to include it. If we want things backwards compatible, you could have a `remove_bias=` kwarg and retain the old behavior by default.

moorepants commented on dtk/process.py:
```diff
@@ -525,9 +532,82 @@ def nextpow2(i):
+        The frequencies where p is a power of 2 close to m.
+    power : ndarray, shape (p,n)
+        The power at each frequency.
+
```
If you can, it would be nice to add these plot directives to show an simple example.

chrismo-schmidt replied:
Done! Please check if it works as expected. I am not experienced with using plot directives.

moorepants commented on dtk/test/test_process.py:
```diff
@@ -433,3 +433,145 @@ def test_time_vector():
+        
+        assert np.allclose(amp, amp_exp), "The result freq_spectrum does n" + \
+                                          "ot match the theoretical expect" + \
+                                          "ation."
```
It is preferable to not use backslashes these days. You can always do:

```python
msg = ("The result freq_spectrum "
      "not match the theoretical "
      "expectation.")
assert np.allclose(amp, amp_exp), msg
```

chrismo-schmidt replied:
Done!

moorepants commented:
Got these two warnings when building the docs:
```
/home/moorepants/src/DynamicistToolKit/dtk/process.py:docstring of dtk.process.cum_pow_spectrum:4: WARNING: Inline substitution_reference start-string without end-string.
/home/moorepants/src/DynamicistToolKit/dtk/process.py:docstring of dtk.process.pow_spectrum:4: WARNING: Inline substitution_reference start-string without end-string.
```

moorepants commented:
Feel free to merge master into this branch. I made a bunch of fixes/changes.

chrismo-schmidt commented:
> Got these two warnings when building the docs:
> 
> ```
> /home/moorepants/src/DynamicistToolKit/dtk/process.py:docstring of dtk.process.cum_pow_spectrum:4: WARNING: Inline substitution_reference start-string without end-string.
> /home/moorepants/src/DynamicistToolKit/dtk/process.py:docstring of dtk.process.pow_spectrum:4: WARNING: Inline substitution_reference start-string without end-string.
> ```

I tried to fix it but not sure if it worked.

chrismo-schmidt commented:
> > An open question is why freq_spectrum removes the bias (f=0)? I think it can be quite crucial to know if a signal as a constant part.
> 
> No idea. I believe I adopted this from a tutorial I found somewhere that explained how this worked. So it is possible it wasn't present or I decided not to include it. If we want things backwards compatible, you could have a `remove_bias=` kwarg and retain the old behavior by default.

I added a kwarg for this. I also made sure that the zero frequency component is not doubled, other then all frequencies >0 to ensure that Parseval's theorem still works on the half spectrum that all the functions return.

moorepants commented on dtk/test/test_process.py:
```diff
@@ -438,3 +438,145 @@ def test_time_vector():
+
+    """
+    
+    def setup(self):
```
For pytest (which I just switched to), this needs to be `def setup_method():`.

moorepants replied:
```suggestion
    def setup_method(self):
```

moorepants commented:
To test the docs locally, do:

```
cd docs\
make html
```

If `make` doesn't work on windows, we will need to solve that.

chrismo-schmidt commented:
I added the sphinx settings and changed the setup method name. 

> To test the docs locally, do:
> 
> ```
> cd docs\
> make html
> ```
> 
> If `make` doesn't work on windows, we will need to solve that.

Indeed, make does not work. I can try to install chocolately somewhen in the coming days. Maybe that helps.

moorepants commented:
There is a `make.bat` file: https://github.com/moorepants/DynamicistToolKit/blob/master/docs/make.bat so, it should work if you call that.

moorepants commented on dtk/process.py:
```diff
@@ -438,17 +438,29 @@ def curve_area_stats(x, y):
 
 
-def freq_spectrum(data, sampleRate):
+def freq_spectrum(data, sample_rate, norm="forward", remove_dc_component=True):
```
```suggestion
def freq_spectrum(data, sampleRate, norm="forward", remove_dc_component=True):
```

We technically shouldn't change the one that is already there because you can call the function as:

```python
freq_spectrum(data=data, sampleRate=sample_rate)
```

but new things we add should use pep8 style for variable names.

moorepants commented on dtk/process.py:
```diff
@@ -506,28 +518,239 @@ def nextpow2(i):
+
+    .. plot::
+       :context: reset
+       :source-include:
```
```suggestion
       :include-source:
```

I wrote in wrong order before.

moorepants commented on dtk/process.py:
```diff
@@ -506,28 +518,239 @@ def nextpow2(i):
+
+    .. plot::
+       :context: reset
+       :source-include:
```
```suggestion
       :include-source:
```

moorepants commented on dtk/process.py:
```diff
@@ -438,17 +438,29 @@ def curve_area_stats(x, y):
         The array of time signals where ``n`` is the number of variables and
         ``m`` is the number of time steps.
-    sampleRate : int
+    sample_rate : int
```
```suggestion
    sampleRate : int
```

moorepants commented on dtk/process.py:
```diff
@@ -506,28 +518,239 @@ def nextpow2(i):
         return n
 
-    time = 1./sampleRate  # sample time
+    time = 1./sample_rate  # sample time
```
```suggestion
    time = 1./sampleRate  # sample time
```

moorepants commented on dtk/process.py:
```diff
@@ -506,28 +518,239 @@ def nextpow2(i):
+    Y = fft(data, n, norm="forward") 
     f = fftfreq(n, d=time)
-    # f = sampleRate/2.*linspace(0, 1, n)
+    # f = sample_rate/2.*linspace(0, 1, n)
```
```suggestion
    # f = sampleRate/2.*linspace(0, 1, n)
```

moorepants commented:
One thing I noticed when pulling this file is that you have trailing whitespaces throughout your work. It's best to set your editor to remove trailing whitespaces and extra lines at the end of the file. I don't know what editor you use, but it should have a setting to do this. Here is what your file looks like in my editor:

![image](https://github.com/moorepants/DynamicistToolKit/assets/276007/aa1d28b1-0256-46f3-9bf1-f1c17fe1fe0a)

moorepants commented:
Your are in!

chrismo-schmidt commented:
> There is a `make.bat` file: https://github.com/moorepants/DynamicistToolKit/blob/master/docs/make.bat so, it should work if you call that.

This worked!
